src/jit/morph.cpp

```diff
@@ -18519,6 +18521,280 @@ void Compiler::fgAddFieldSeqForZeroOffset(GenTree* op1, FieldSeqNode* fieldSeq)
+        DoPostOrder       = true,
+        ComputeStack      = true,
+        DoLclVarsOnly     = false,
+        UseExecutionOrder = false,
+    };
+
+    LocalAddressVisitor(Compiler* comp)
+        : GenTreeVisitor<LocalAddressVisitor>(comp), m_valueStack(comp->getAllocator(CMK_IndirAssignMap))
+    {
+    }
+
+    void VisitStmt(GenTreeStmt* stmt)
+    {
+        WalkTree(&stmt->gtStmtExpr, nullptr);
+        assert(m_valueStack.Height() == 1);
+        m_valueStack.Reset();
+    }
+
+    fgWalkResult PreOrderVisit(GenTree** use, GenTree* user)
+    {
+        GenTree* node = *use;
+
+        if (node->OperIs(GT_FIELD))
+        {
+            MorphStructField(node, user);
+        }
+        else if (node->OperIs(GT_LCL_VAR, GT_LCL_FLD))
+        {
+            unsigned lclNum = node->AsLclVarCommon()->GetLclNum();
+
+            if (m_compiler->lvaIsImplicitByRefLocal(lclNum))
+            {
+                // Keep track of the number of appearances of each promoted implicit
+                // byref (here during address-exposed analysis); fgMakeOutgoingStructArgCopy
+                // checks the ref counts for implicit byref params when deciding if it's legal
+                // to elide certain copies of them.
+                LclVarDsc* varDsc = &m_compiler->lvaTable[lclNum];
+                JITDUMP("Incrementing ref count from %d to %d for V%02d in %s\n", varDsc->lvRefCnt(RCS_EARLY),
+                        varDsc->lvRefCnt(RCS_EARLY) + 1, lclNum, __FUNCTION__);
+
+                varDsc->incLvRefCnt(1, RCS_EARLY);
+            }
+
+            if (node->OperIs(GT_LCL_FLD))
+            {
+                MorphLocalField(node, user);
+            }
+        }
+
+        PushValue(node);
+
+        return Compiler::WALK_CONTINUE;
+    }
+
+    fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+    {
+        GenTree* node = *use;
+
+        switch (node->OperGet())
+        {
+            case GT_LCL_VAR:
+                assert(TopValue(0).Node() == node);
+                TopValue(0).Location(node->AsLclVar()->GetLclNum());
+                break;
+
+            case GT_LCL_FLD:
+                assert(TopValue(0).Node() == node);
+                TopValue(0).Location(node->AsLclFld()->GetLclNum(), node->AsLclFld()->gtLclOffs);
+                break;
+
+            case GT_ADDR:
+                assert(TopValue(1).Node() == node);
+                TopValue(1).Address(TopValue(0));
+                PopValue();
+                break;
+
+            case GT_FIELD:
+                if (node->AsField()->gtFldObj != nullptr)
+                {
+                    assert(TopValue(1).Node() == node);
+                    TopValue(1).Field(TopValue(0), node->AsField()->gtFldOffset);
+                    PopValue();
+                }
+                else
+                {
+                    assert(TopValue(0).Node() == node);
+                }
+                break;
+
+            case GT_OBJ:
```

Comment:
Unlike for `GT_IND`, the old code did pay attention to the size of `GT_OBJ`/`BLK`. Need to restore that. Interesting that this doesn't have any bad observable side effects.

Reply:
Added size checking. Though not right here but later, when an indir node is actually consumed.